uaclient/clouds/identity.py

```diff
@@ -0,0 +1,50 @@
+import json
+
+from uaclient import exceptions
+from uaclient import clouds
+from uaclient import status
+from uaclient import util
+
+CLOUDINIT_RESULT_FILE = "/var/lib/cloud/data/result.json"
+
+
+@util.retry(FileNotFoundError, [1, 2])
+def get_cloud_type_from_result_file(result_file=CLOUDINIT_RESULT_FILE) -> str:
+    result = json.loads((util.load_file(result_file)))
+    dsname = result["v1"]["datasource"].split()[0].lower()
+    return dsname.replace("datasource", "")
+
+
+def get_cloud_type() -> str:
+    if util.which("cloud-id"):
+        # Present in cloud-init on >= Xenial
+        out, _err = util.subp(["cloud-id"])
+        return out.strip()
+    try:
+        return get_cloud_type_from_result_file()
+    except FileNotFoundError:
+        pass
+    return ""
+
+
+def cloud_instance_factory() -> clouds.UAPremiumCloudInstance:
+    from uaclient.clouds import aws
+
+    cloud_instance_map = {
+        "aws": aws.UAPremiumAWSInstance,
```

Comment:
I think we should normalise these in `get_cloud_type`, that's already having to handle the differences between trusty and later releases.

Reply:
good point. +1